FriCAS spkg-configure and Feature

dimpase opened the pull request:
As FriCAS is only used as an executable, this is straightforward; quite a number of systems has Fricas 1.3.8, so this is useful, too.

The FriCAS pexpect interface now uses the new `Executable` feature to determine the absolute pathname of the fricas executable (unless executed remotely). 
This is made possible by a simple refactor of the `sage.interfaces.expect.Expect` class: Computing the effective command line is no longer done in `set_server_and_command` (called by `__init__`); it is delayed until an interface is started and needs the command line.

- Fixes #35837 
- Fixes #33575

Dependencies: The changes in `sage.interfaces` outside of `.expect` and `.fricas` are all from the following PR and do not need review.

- Depends on #36656 (merged here)

Commit:
Add # optional - numpy etc.

Commit:
src/sage/interfaces/tides.py: pycodestyle fix

Commit:
Massive modularization fixes

Commit:
More # optional

Commit:
sage.interfaces: Modularization fixes

Commit:
More # optional

Commit:
More # optional

Commit:
More # optional

Commit:
More # optional

dimpase committed:
more distros for fricas

dimpase committed:
spkg-configure for fricas

dimpase committed:
remove debugging echo

kiwifb commented on build/pkgs/fricas/distros/gentoo.txt:
```diff
@@ -0,0 +1 @@
+sci-mathematics/fricas[sbcl]
```
Why do you impose it to be build with `sbcl`. Sure it is probably the fastest lisp around, but is there a specific reason that you need to have it?

dimpase replied:
ok

dimpase committed:
do not insist on sbcl

dimpase committed:
modify the way to send )quit to fricas

This is needed as SBCL-built FriCAS does not output anything
if )quit is passed via -eval option, as opposed to ECL-built FriCAS.

mkoeppe commented on build/pkgs/fricas/spkg-configure.m4:
```diff
@@ -0,0 +1,16 @@
+SAGE_SPKG_CONFIGURE(
+    [fricas], [
+        AC_CACHE_CHECK([for FriCAS >= 1.3.8], [ac_cv_path_FRICAS], [
```
the indentation is a bit wild

dimpase committed:
updated design of feature

mkoeppe commented on src/sage/features/fricas.py:
```diff
@@ -0,0 +1,65 @@
+        """
+        command = ["fricas -nosman", "2+2"]
+        try:
+            lines = subprocess.check_output(command, stderr=subprocess.STDOUT)
```
if you don't use `shell=True`, this cannot work

dimpase replied:
I still don't understand what we are doing here. FriCAS feature is something configured; the testing framework should **test the configuration**, and do  not engage in these run-time double-checks.

As soon as testing framework knows that FriCAS should be available, it must just ignore `# optional - fricas` tags,
and treat FriCAS as available.

I don't get why distributions have any need for these, either.

dimpase committed:
better string to match (for various lisps)

Commit:
./sage -fixdoctests --no-test src/sage/interfaces

Commit:
sage.interfaces: Update # needs

Commit:
src/sage/interfaces: Update file-level doctest tag

Commit:
src/sage/interfaces: sage -fixdoctests --only-tags

Commit:
Fix # needs for sagemath-pari

Commit:
Add # needs

Commit:
Update # needs

Commit:
src/sage/interfaces/fricas.py: Use file-level doctest tag

Commit:
sage.interfaces.expect: Delay construction of command to first use, handle Executable

Commit:
src/sage/interfaces/fricas.py: Use the FriCAS feature less eagerly

Commit:
sage.interfaces.expect: Fix up ulimit

Commit:
sage.interfaces.expect: Update doctest output

Commit:
src/sage/interfaces/fricas.py: Do not rely on '# tol' for complex tolerances; cf. #36631

Commit:
src/sage/functions/special.py: Use 'abs tol' instead of 'tol' in two examples
